lib/fibonacci.rb

```diff
@@ -1,9 +1,41 @@
 # Improved Fibonacci
-
-# Time Complexity - ?
-# Space Complexity - ?  (should be O(n))
 # Hint, you may want a recursive helper method
 # Hint:  Fib(0) = 0, Fib(1) = 1, work up from there
+
+# Time Complexity - O(n)
+# >> max n stack frames at a time (one fib_helper branch is explored at a time)
+# >> hash accessing is O(1)
+#
+# Space Complexity - O(n)
+# >> max n stack frames at a time
+# >> stored_fibs hash of up to n values
+
 def fibonacci(n)
```

Comment:
This works and it's O(n) for both space and time complexity, however can you see a way to do it without needing a hash of all the numbers from 0..n?